### PR TITLE
Performance guide is updated to mention profiling via Build Scans as …

### PR DESCRIPTION
**Pull request by timyates**
…a first-class citizen

From the [Trello card for CS](https://trello.com/c/DKxkoogj/1592-performance-guide-is-updated-to-mention-profiling-via-build-scans-as-a-first-class-citizen):

> ## Motivation
>
> The current Performance guide does not show the new features and deeper insight that can be gained by using Build Scans.

Updated several sections to show using build scans to diagnose common performance issues

### CODE REVIEW

File: contents/index.adoc [hunk 1]
```diff
@@ -147,12 +147,32 @@ The rest of the summary relates to the main categories, which we cover in detail
...
+
+Hopefully you can see how the ability to drill into the data provided by build scans can really help in optimizing a build.
+
+The next few subsections introduce techniques that can help improve the configuration time and explain why they work.
 
 ### Apply plugins judiciously
 
 Every plugin that you apply to a project adds to the overall configuration time. Some plugins have a greater impact than others. That doesn’t mean you should avoid using plugins, but you should take care to only apply them where they’re needed. For example, it’s easy to apply plugins to all projects via `allprojects {}` or `subprojects {}` even if not every project needs them.
 
+In our above build scan example, we can see that `script-a.gradle` is applied to 3 projects inside the build:
```
Comment (facewindu) [inline contents/index.adoc]: this section is about plugins. It would be nice to have an example with a plugin and not a script.

File: contents/index.adoc [hunk 2]
```diff
@@ -147,12 +147,32 @@ The rest of the summary relates to the main categories, which we cover in detail
...
+
+image::script-a-application.png[title="Showing the application of script-a to our build"]
+
+This script takes a second to run, and we can see it is applied to 3 projects from the root `build.gradle` file, and so is introducing a 3 second delay to our configuration.
+
 Ideally, plugins should not incur a significant configuration-time cost. If they do, the focus should be on improving the plugin. Nonetheless, in projects with many modules and a significant configuration time, you should spend a little time identifying any plugins that have a notable impact. The only reliable way to do this is by running a build twice: once with the plugin applied and once without.
```
Comment (facewindu) [inline contents/index.adoc]: "The only reliable way to do this is by running a build twice: once with the plugin applied and once without."
this is now false. You can see it via build scans directly.

File: contents/index.adoc [hunk 3]
```diff
@@ -147,12 +147,32 @@ The rest of the summary relates to the main categories, which we cover in detail
 
 As the user guide describes in https://docs.gradle.org/current/userguide/build_lifecycle.html[the build lifecycle chapter], a Gradle build goes through three phases: initialization, configuration, and execution. The important thing to understand here is that configuration code always executes regardless of which tasks will run. That means any expensive work performed during configuration will slow every invocation down, even simple ones like `gradle help` and `gradle tasks`.
 
-The profile report will help you identify which projects take the most time to configure, but that’s all. The next few subsections introduce techniques that can help improve the configuration time and explain why they work.
+A profile report will help you identify which projects take the most time to configure, but that’s all. The advantage of using a build scan is that it lets you drill down into this configuration time, and see what is causing any slowness.
+
+In the above xref:build-scan-performance[] we can see that Configuration is taking over 14 seconds. Clicking on the _"Configuration"_&nbsp; tab at the top of the page will break this down for us.
```
[the remark this answers is not in the document]
Comment (facewindu) [reply contents/index.adoc]: maybe that's a bigger change and we should discuss first, as I see it was already the case in this document.

File: contents/index.adoc [hunk 4]
```diff
@@ -209,7 +229,13 @@ include::{samplescodedir}/copy-3.gradle[tags=copy]
 
 Note that the `from()` declaration doesn’t resolve the dependencies because you’re using the https://docs.gradle.org/current/userguide/artifact_dependencies_tutorial.html#configurations[dependency configuration] itself as an argument, not its files. The `Copy` task handles the resolution of the configuration itself during task execution, which is exactly what you want.
 
-The performance page of build scans explicitly shows how dependency resolution time is split across project configuration and task execution, so it's easy to identify this particular issue. If you're using the older profile reports, a simple way to determine whether you’re resolving dependencies during configuration is to run
+The performance page of build scans explicitly shows how dependency resolution time is split across project configuration and task execution:
+
+image::bad-dependency-resolution.png[title="Bad dependency resolution at configuration time."]
+
+Here, it's easy to identify this particular issue; we should have a "Project configuration" time of zero, and this example shows we seem to be resolving dependencies too early in the lifecycle.
```
Comment (facewindu) [inline contents/index.adoc]: maybe rephrase to say that it we should have 'resolved dependencies during project configuration time' of zero, not " "Project configuration" time of zero", since it's not possible.

Comment (facewindu) [reply contents/index.adoc]: also mention here that the settings and suggestions tab will tell which dependencies are resolved at configuration time.
The screenshot also shows it, but it's misleading, because it shows all resolved configuration (you would also see the ones during task execution)

File: contents/index.adoc [hunk 5]
```diff
@@ -272,6 +304,16 @@ It is also possible to use a shared build cache service, like https://gradle.com
 
 We have a complete guide about https://guides.gradle.org/using-build-cache/[using the build cache]. It covers the different scenarios caching can improve, and detailed discussions of the different caveats you need to be aware of when enabling caching for a build.
 
+Again, build scans can help you investigate how well your tasks are caching. In the performance screen, we have a tab titled _"Build cache"_:
```
Comment (facewindu) [inline contents/index.adoc]: not 'we'

Comment (facewindu) [inline contents/index.adoc]: shoulw we mention it's not always displayed (and the condition to have the tab displayed)?

File: contents/index.adoc [hunk 6]
```diff
@@ -272,6 +304,16 @@ It is also possible to use a shared build cache service, like https://gradle.com
...
+
+image::cache-performance.png[title="Inspecting the performance of the build cache for a build."]
+
+This shows you statistics about how many tasks were fetched from a cache, which cache was used, along with download and unpack rates for these cache entries.
```
Comment (facewindu) [inline contents/index.adoc]: were fetched and stored
when Craig will have finished his in progress work

Maybe this deserves to wait a little bit to have the updated screenshot

Comment (facewindu) [inline contents/index.adoc]: download and unpack rates, but also upload and pack rates

File: contents/index.adoc [hunk 7]
```diff
@@ -285,6 +327,9 @@ Gradle also supports other forms of partial build via the _base_ plugin, which a
 
 These tasks are slower than just running `build` in the target project as they do more work, but they are an effective alternative to running `gradle build`, which runs `build` in _all_ the projects of a multi-project build.
 
+### Using build scans to trace un-cached tasks
```
Comment (facewindu) [inline contents/index.adoc]: this section is missing content

Comment (timyates) [reply contents/index.adoc]: Gone

File: contents/index.adoc (hunk 6 above)
Comment (timyates) [reply contents/index.adoc]: Fixed

File: contents/index.adoc (hunk 5 above)
Comment (timyates) [reply contents/index.adoc]: Fixed

File: contents/index.adoc (hunk 4 above)
Comment (timyates) [reply contents/index.adoc]: Fixed

File: contents/index.adoc (hunk 2 above)
Comment (timyates) [reply contents/index.adoc]: Removed sentence

File: contents/index.adoc (hunk 1 above)
Comment (timyates) [reply contents/index.adoc]: Do we have one in the GE Training?  Or am I going to need to make one up?

File: contents/index.adoc [hunk 8]
```diff
@@ -147,12 +147,32 @@ The rest of the summary relates to the main categories, which we cover in detail
...
 
-The profile report will help you identify which projects take the most time to configure, but that’s all. The next few subsections introduce techniques that can help improve the configuration time and explain why they work.
+A profile report will help you identify which projects take the most time to configure, but that’s all. The advantage of using a build scan is that it lets you drill down into this configuration time, and see what is causing any slowness.
+
+In the above xref:build-scan-performance[] we can see that Configuration is taking over 14 seconds. Clicking on the _"Configuration"_&nbsp; tab at the top of the page will break this down for us.
+
+image::build-scan-configuration-breakdown.png[title="Build scan configuration breakdown"]
+
+Here we can see the scripts and plugins that were applied to our project in descending order of how long they took to apply. If we take the top offender, `script-b.gradle`, we can expand it up to see where this script is applied to our build
```
[the remark this answers is not in the document]
Comment (timyates) [reply contents/index.adoc]: I think you can have a comma after `Here` (above) or before `in descending`, but not both as it suggests the text between the commas is describing the outer sentence...

[the remark this answers is not in the document]
Comment (timyates) [reply contents/index.adoc]: Fixed

[the remark this answers is not in the document]
Comment (timyates) [reply contents/index.adoc]: Fixed

[the remark this answers is not in the document]
Comment (timyates) [reply contents/index.adoc]: Tried to de-humanize the script

File: contents/index.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (timyates) [reply contents/index.adoc]: I prefer guides to hold my hand and help me through a task as a (admittedly one man) team.

I guess this needs wider discussion

[the remark this answers is not in the document]
Comment (timyates) [reply contents/index.adoc]: Fixed

File: contents/index.adoc (hunk 5 above)
Comment (timyates) [reply contents/index.adoc]: What is the condition?  That you're using the build cache?

File: contents/index.adoc (hunk 6 above)
Comment (timyates) [reply contents/index.adoc]: Not sure what to do here...

File: contents/index.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (facewindu) [reply contents/index.adoc]: It can help you and "hold your hand" without saying "we", like we're the best friends in the world.

File: contents/index.adoc (hunk 1 above)
Comment (facewindu) [reply contents/index.adoc]: I don't think we have.

File: contents/index.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (facewindu) [reply contents/index.adoc]: Thx

File: contents/index.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (facewindu) [reply contents/index.adoc]: Thx

[the remark this answers is not in the document]
Comment (facewindu) [reply contents/index.adoc]: Thx

File: contents/index.adoc (hunk 2 above)
Comment (facewindu) [reply contents/index.adoc]: Thx

File: contents/index.adoc (hunk 4 above)
Comment (facewindu) [reply contents/index.adoc]: Thx

File: contents/index.adoc (hunk 5 above)
Comment (facewindu) [reply contents/index.adoc]: Thx

Comment (facewindu) [reply contents/index.adoc]: gradle 4.0+, plugin 1.8+, build cache globally enabled for the build

File: contents/index.adoc (hunk 6 above)
Comment (facewindu) [reply contents/index.adoc]: "were fetched from a cache" => "were fetched and stored from/to a cache (or sth like that)
Soon, you will have a screenshot that can show this added statistics. Ask @craigatk

Comment (facewindu) [reply contents/index.adoc]: thx

File: contents/index.adoc (hunk 7 above)
Comment (facewindu) [reply contents/index.adoc]: why, isn't it a nice section to have in this guide ?

File: contents/index.adoc (hunk 1 above)
Comment (timyates) [reply contents/index.adoc]: Posted another commit which uses plugins `com.example.foo` and `com.example.bar` instead of then scripts for this bit of the doc

File: contents/index.adoc (hunk 7 above)
Comment (timyates) [reply contents/index.adoc]: It was moved to the

### Caching task outputs

section (line 298ish)

Comment (timyates) [reply contents/index.adoc]: This was a placeholder, I forgot to delete 😿

**Comment by timyates** [comment]
Updated to use plugins instead of scripts (for the config time section)

Might burn this PR down and start again if @facewindu agrees?  Github reviews are apparently a nightmare 😨

**Comment by timyates** [comment]
Thanks to @breskeby we now have a crucible review over here

https://code-review.gradle.org/cru/GGP-1
